api/proto/config/storage.proto

```diff
@@ -242,3 +242,39 @@ message DisksConfig {
   DisksArrayType array_type = 2;
   repeated DisksConfig children = 3;
 }
+
+// RollbackCmd - snapshot rollback command
+message RollbackCmd {
+  string snpshot_uuid = 1;
```

Comment:
Typo snapshot_uuid I assume.

Reply:
Also, is this the uuid field in the SnapshotConfig? Some more comments would be helpful.

I can see how the controller requests one or more snapshots (uuid=1 for volume_uuid=X, then add a second SnapshotCmd with uuid=2 for volume_uuid=X) and then later it can send a RollbackCmd with snapshot_uuid=1? 

Is that correct?

If so can you please confirm that the SnapshotConfig will have uuid=1 (since we want to keep that snapshot around) and the RollbackCmd have snapshot_uuid=1?

Basically what you call snapshot_uuid is the uuid to which we want to roll back - the target of the rollback?

Reply:
> I can see how the controller requests one or more snapshots (uuid=1 for volume_uuid=X, then add a second SnapshotCmd with uuid=2 for volume_uuid=X) and then later it can send a RollbackCmd with snapshot_uuid=1?
> 
> Is that correct?

It is assumed that the RollbackCmd message cannot come separately from the configuration, and will always be part of the SnapshotConfig for every snapshot.
If I understand correctly, then we will always receive from the controller a complete list of SnapshotConfig for all snapshots that the controller knows about.
Based on your example, if I understand correctly,
then after the snapshot with uuid=2 has been created,
we can also get a SnapshotConfig for uuid=1, where the counter in RollbackCmd will be +1. Thus, we will roll back to the state of the logical volume at the time of the snapshot with uuid=1, and the snapshot with uuid=2 will be marked as implicitly deleted (If the controller did not take care of deleting the snapshot with uuid=2 before the rollback command).

> If so can you please confirm that the SnapshotConfig will have uuid=1 (since we want to keep that snapshot around) and the RollbackCmd have snapshot_uuid=1?

The fields from RollbackCmd (snapshot_uuid, volume_uuid) will always match the fields from the SnapshotConfig (uuid, volume_uuid) to which RollbackCmd belongs.

> Basically what you call snapshot_uuid is the uuid to which we want to roll back - the target of the rollback?

snapshot_uuid is the uuid from the SnapshotConfig and the real and unique snapshot name in ZFS. So yes, that's the purpose of the rollback for volume_uuid


If I misunderstood you, please correct me and we will deal with this in more detail.

Reply:
How about calling the field snapshot_uuid in both places i.e., also in the SnapshotConfig? That makes the relationship more clear.

Reply:
Also, it probably makes sense capturing the above discussion in a comment in the API for the RollbackCmd.

Reply:
Corrected and added a comment